### DISCUSSION TITLE
`Bitmap` does not save into a memory stream

### DISCUSSION THREAD

**Discussion by Badwolf369** [AvaloniaUI/Avalonia#14490] [Q&A] [answered]
## Describe the bug
When trying to use the `Save()` method on `Media.Imaging.Bitmap` to save the image into a memory stream, the memory stream remains unchanged.

## To Reproduce
1. Create a new `Media.Imaging.Bitmap` from a file and open a memory stream.
2. Pause using debugger. Inspect memory stream and take note that it is empty and writeable.
3. Save the bitmap into the memory stream using `Bitmap.Save()`.
4. Pause and inspect the memory stream again and observe that it is unchanged.
``` C#
Bitmap bitmap = new Bitmap("image.png");
using (MemoryStream imageStream = new MemoryStream())   // Paused here with debugger
{
    bitmap.Save(imageStream); // After stepping past this line, imageStream remains unchanged
}
```

## Expected behavior
I should be able to observe on the debugger that after `Bitmap.Save()` is called, the image stream now has a length of around 900 instead of 0.

## Screenshots
Empty memory stream immediately after initialization
![image](https://github.com/AvaloniaUI/Avalonia/assets/22065391/51b6adf5-a8e9-44f5-b859-e7a0f82a89b1)
Memory stream after calling `Bitmap.Save()`
![image](https://github.com/AvaloniaUI/Avalonia/assets/22065391/9de68bd2-f701-4ed5-b74a-a9ff71051149)
Memory stream when the exact same image is written to it using a different image object, specifically an OpenCV Mat object
![image](https://github.com/AvaloniaUI/Avalonia/assets/22065391/5a3ef2d1-12cb-42a7-a11a-9e76e6667cde)

## Environment 
- OS: Windows 10
- Avalonia-Version: 11.0.2

## Additional context
The method where this is happening exists in a proper Avalonia app, however the code is being executed headlessly in XUnit if that has any relevance.

**Comment by stevemonaco** [comment]
I've tried to reproduce this in an Avalonia desktop app and could not with the following code:

```cs
private void Button_Click(object? sender, RoutedEventArgs e)
{
    Bitmap bitmap = new Bitmap("Resources/image.png");

    using (MemoryStream imageStream = new MemoryStream())
    {
        bitmap.Save(imageStream);
        imageStream.Seek(0, SeekOrigin.Begin);
        var newBitmap = new Bitmap(imageStream);
        image.Source = newBitmap;
    }
}
```

The image displays as expected and the debugger shows the `MemoryStream` has content. Worked on both 11.0.2 and nightly.

![image](https://github.com/AvaloniaUI/Avalonia/assets/1782158/e8744f1f-83c9-4cfe-8de0-84f3dabb54d4)

**Comment by maxkatz6** [comment, marked as answer]
> however the code is being executed headlessly in XUnit if that has any relevance.

Have you enabled skia drawing?

**Reply by Badwolf369** [reply]
This fixed the issue when creating a Bitmap from a file, but throws `System.ArgumentNullException : Value cannot be null. (Parameter 'data') at SkiaSharp.SKBitmap.Decode(SKData data)` when creating a Bitmap from a memory stream

**Reply by maxkatz6** [reply]
Make sure that Stream position is at 0.

**Reply by Badwolf369** [reply]
That fixed the issue, thankyou! It would probably be a good idea to mention something about this in the "Testing with XUnit" Documentation

**Reply by timunie** [reply]
a PR is welcome

**Reply by maxkatz6** [reply]
Possibly as a reminder. It's a pretty common mistake with MemoryStream.

**Comment by timunie** [comment]
@Badwolf369 please update to 11.0.7 and if issue persists, please upload a minimum sample

**Comment by timunie** [comment]
given that we cannot reproduce it on our side, I'll make a discussion from it. If we find out it is an actually reproducible bug, I'll reopen this issue.

**Comment by Badwolf369** [comment]
I have done some further testing and upgraded to 11.0.7 to find when and how exactly the issue is appearing.
The Bitmap is empty when created from either a file or a memory stream in headless XUnit testing, but it is created fine in normal program running. The Bitmap being created empty is what makes it *appear* like it is not being saved to memory, when really there was nothing to save in the first place.
